File: api-go/internal/middleware/auth_middleware.go
// Package middleware provides JWT authentication middleware for Fiber.
package middleware

import (
	"strings"

	"challenge-api-go/internal/models"

	"github.com/gofiber/fiber/v2"
	"github.com/golang-jwt/jwt/v5"
)

// JWTMiddleware creates a Fiber middleware that validates JWT tokens.
func JWTMiddleware(secret string) func(*fiber.Ctx) error {
	return func(c *fiber.Ctx) error {
		authHeader := c.Get("Authorization")
		if authHeader == "" {
			return c.Status(fiber.StatusUnauthorized).JSON(models.APIResponse{
				Success: false,
				Error: &models.APIError{
					Code:    "UNAUTHORIZED",
					Message: "Falta el header de autorización",
				},
			})
		}

		parts := strings.SplitN(authHeader, " ", 2)
		if len(parts) != 2 || parts[0] != "Bearer" {
			return c.Status(fiber.StatusUnauthorized).JSON(models.APIResponse{
				Success: false,
				Error: &models.APIError{
					Code:    "UNAUTHORIZED",
					Message: "Formato de autorización inválido. Use: Bearer <token>",
				},
			})
		}

		tokenStr := parts[1]
		token, err := jwt.Parse(tokenStr, func(token *jwt.Token) (interface{}, error) {
			if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
				return nil, jwt.ErrSignatureInvalid
			}
			return []byte(secret), nil
		})

		if err != nil || !token.Valid {
			return c.Status(fiber.StatusForbidden).JSON(models.APIResponse{
				Success: false,
				Error: &models.APIError{
					Code:    "FORBIDDEN",
					Message: "Token inválido o expirado",
				},
			})
		}

		// Store the raw token for forwarding to downstream services.
		c.Locals("token", tokenStr)
		return c.Next()
	}
}

File: api-node/src/middleware/auth.middleware.js
// JWT authentication middleware for Express.
const jwt = require("jsonwebtoken");

const JWT_SECRET = process.env.JWT_SECRET || "dev_secret_change_me";

function authMiddleware(req, res, next) {
  const authHeader = req.headers.authorization;
  if (!authHeader) {
    return res.status(401).json({
      success: false,
      error: { code: "UNAUTHORIZED", message: "Falta el header de autorización" },
    });
  }

  const parts = authHeader.split(" ");
  if (parts.length !== 2 || parts[0] !== "Bearer") {
    return res.status(401).json({
      success: false,
      error: {
        code: "UNAUTHORIZED",
        message: "Formato de autorización inválido. Use: Bearer <token>",
      },
    });
  }

  try {
    const decoded = jwt.verify(parts[1], JWT_SECRET);
    req.user = decoded;
    req.token = parts[1];
    next();
  } catch (err) {
    return res.status(403).json({
      success: false,
      error: { code: "FORBIDDEN", message: "Token inválido o expirado" },
    });
  }
}

module.exports = authMiddleware;
